# 120 years of Olympic history: athletes and results

## 2025-26 Group: 2

*This is the MN5813 Group Assignment from Group 2.*

- Student 1 : 101114910
- Student 2 : 101116129
- Student 3 : 101115762

## Project summary

In [1]:
# Add your project summary here [app. 200-250 words]

## Generative AI

##### Add your statement / declaration here [be as specific as possible]
We used AI to help us with the following:

- what relative path is in order to share and work with our working files and collaborate effectivily as a team 

## Introduction

# Add your introduction here [app. 200-300 words]
Idea - to tell a bit of Olympic History

About Dataset
Context
This is a historical dataset on the modern Olympic Games, including all the Games from Athens 1896 to Rio 2016. 

Note that the Winter and Summer Games were held in the same year up until 1992. After that, they staggered them such that Winter Games occur on a four year cycle starting with 1994, then Summer in 1996, then Winter in 1998, and so on. A common mistake people make when analyzing this data is to assume that the Summer and Winter Games have always been staggered.


# Formulate project aim & set of objectives

### **Project aim:** To explore the data set to acquire knowledge and enhance practical skills in the Python programming language

### **Project Objectives:**

*Objectives are subject to change*

1st objective: Understand the data
In order to achieve this objective we want to measure the following:
  - Calculate the average age of athletes for each Olympic event
  - Determine the top 10 countries with the most gold medals
  - Find the athlete with the most medals in each sport

additional
(are there any countries that have never won Olympic medals ?)
    
2nd objective: Visually explore the data set
In ordet to achieve this objective we want to create two visualisations representing the outcome of our data analysis

3rd objective: Make recommendations as for what else can be done 


## Setup

### First we need to import necessary libraries

In [29]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns        # - for to visualise results across three dimensions (hopefully)

# Set display options for better output
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)
pd.set_option("display.width", None)

### Load data

data_source = "https://www.kaggle.com/heesoo37/120-years-of-olympic-history-athletes-and-results"        # Add URL to raw data here

[data source](https://www.kaggle.com/heesoo37/120-years-of-olympic-history-athletes-and-results)

Zip file is downloaded from Kaggle to our laptops. 

We extract zip file and find two csv files in this folder:
- athlete_events
- noc_regions

### Convert into DataFrame

We use `pd.read_csv()` to convert data from both files found in Olympic History folder into Pandas DataFrames. 

And we use relative paths to the data for consistency. 

First We want to know how many rows and columns in our both data sets and names of the columns. 

We use `len(data_frame)` to find how many rows in our data sets; `len(data_frame.columns)` to find the number of columns and `list(data_frame.columns)` to list names of all the columns in our data sets.

In [49]:
athlete_events = pd.read_csv("data/athlete_events.csv")
noc_regions = pd.read_csv("data/noc_regions.csv")

print(f"Dataset 'athlete_events' loaded: {len(athlete_events)} rows, {len(athlete_events.columns)} columns")
print(f"\nColumn names:\n{list(athlete_events.columns)}")

print(f"\n\nDataset 'noc_regions' loaded: {len(noc_regions)} rows, {len(noc_regions.columns)} columns")
print(f"\nColumn names:\n{list(noc_regions.columns)}")

Dataset 'athlete_events' loaded: 271116 rows, 15 columns

Column names:
['ID', 'Name', 'Sex', 'Age', 'Height', 'Weight', 'Team', 'NOC', 'Games', 'Year', 'Season', 'City', 'Sport', 'Event', 'Medal']


Dataset 'noc_regions' loaded: 230 rows, 3 columns

Column names:
['NOC', 'region', 'notes']


The first few rows with column names of `athlete_events` and `noc_regions` DataFrames we have created from `data_source` are available in the Appendices A and B - to see how raw data looks like. 

The first data set `athlete_events` holds all the main information in 15 columns and 271116 rows.

The second data set `noc_regions` has 3 colomns and 230 rows. It provides full names of the countries and some notes about these countries.  

We inentified that `athlete_events` and `noc_regions` both have NOC column (short name of the country in three letters). This column can be a key to merge two data sets. 

### Merging our data from two data frames a single DataFrame

We can now join the two dataframes using as a key the NOC column with Pandas `Merge` function. 

In [36]:
df = pd.merge(athlete_events, noc_regions, on="NOC", how="left")

### Initial exploration

We start by exploring our data to understand its structure, contents, and quality.

We display the first 5 rows of our merged data frame with `.head()` function.

Now we can see columns content and all column names. 

In [38]:
print("First 5 rows:")
df.head()

First 5 rows:


ID                      Name Sex   Age  Height  Weight            Team  \
0   1                 A Dijiang   M  24.0   180.0    80.0           China   
1   2                  A Lamusi   M  23.0   170.0    60.0           China   
2   3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN         Denmark   
3   4      Edgar Lindenau Aabye   M  34.0     NaN     NaN  Denmark/Sweden   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   

   NOC        Games  Year  Season       City          Sport  \
0  CHN  1992 Summer  1992  Summer  Barcelona     Basketball   
1  CHN  2012 Summer  2012  Summer     London           Judo   
2  DEN  1920 Summer  1920  Summer  Antwerpen       Football   
3  DEN  1900 Summer  1900  Summer      Paris     Tug-Of-War   
4  NED  1988 Winter  1988  Winter    Calgary  Speed Skating   

                              Event Medal       region notes  
0       Basketball Men's Basketball   NaN        China   NaN  
1      Judo Men's Extra-Lightweight   NaN        China   NaN  
2           Football Men's Football   NaN      Denmark   NaN  
3       Tug-Of-War Men's Tug-Of-War  Gold      Denmark   NaN  
4  Speed Skating Women's 500 metres   NaN  Netherlands   NaN

We use `info` function to see descriptive general information about our data frame: datatypes, column names, number of columns, number of entries in each column and other information.


In [50]:
print("Dataset information:")
df.info()

Dataset information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271116 entries, 0 to 271115
Data columns (total 17 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   ID      271116 non-null  int64  
 1   Name    271116 non-null  object 
 2   Sex     271116 non-null  object 
 3   Age     261642 non-null  float64
 4   Height  210945 non-null  float64
 5   Weight  208241 non-null  float64
 6   Team    271116 non-null  object 
 7   NOC     271116 non-null  object 
 8   Games   271116 non-null  object 
 9   Year    271116 non-null  int64  
 10  Season  271116 non-null  object 
 11  City    271116 non-null  object 
 12  Sport   271116 non-null  object 
 13  Event   271116 non-null  object 
 14  Medal   39783 non-null   object 
 15  region  270746 non-null  object 
 16  notes   5039 non-null    object 
dtypes: float64(3), int64(2), object(12)
memory usage: 35.2+ MB


Now we use the `describe` function to view statistical summary of numerical columns. 

Wow, we see that minimum age of athlets is 10 years and maximum is 97 years. It's becoming interesting. 

In [40]:
print("Summary statistics:")
df.describe()

Summary statistics:


ID            Age         Height         Weight  \
count  271116.000000  261642.000000  210945.000000  208241.000000   
mean    68248.954396      25.556898     175.338970      70.702393   
std     39022.286345       6.393561      10.518462      14.348020   
min         1.000000      10.000000     127.000000      25.000000   
25%     34643.000000      21.000000     168.000000      60.000000   
50%     68205.000000      24.000000     175.000000      70.000000   
75%    102097.250000      28.000000     183.000000      79.000000   
max    135571.000000      97.000000     226.000000     214.000000   

                Year  
count  271116.000000  
mean     1978.378480  
std        29.877632  
min      1896.000000  
25%      1960.000000  
50%      1988.000000  
75%      2002.000000  
max      2016.000000

**Observations from initial exploration**:

1. **Data structure**: 271116 rows × 17 columns
2. **Data types**: Mix of numeric (int64, float64) and text (object)
3. **Missing values**: Some columns have missing data (Age, Height, Weight, Medal, region, notes)
4. **Value ranges**: Numeric values appear generally reasonable (75% of athlets is 28 years old or younger, with maximum height 183 and weight 79kg). Though there are some outliers there (age: 10-97, height: 127-226, weight: 25-214). Some athlets are very young (10-12 years old) and some weigh only 25 kg. 

## Data management

### Data cleaning
Data cleaning is essential before analysis. We'll identify and handle:

- Missing values
- Data type issues
- Duplicates
- Outliers (if needed)

#### Section 1: Missing values

First, we identify which columns have missing data and how much.

In [55]:
# TODO: Handle missing values

# Count missing values
missing = df.isnull().sum()
missing_pct = (df.isnull().sum() / len(df) * 100).round(1)

missing_summary = pd.DataFrame({
    "Missing_Count": missing,
    "Missing_Percent": missing_pct
})

print("Missing values summary:")
print(missing_summary[missing_summary['Missing_Count'] > 0].sort_values("Missing_Count", ascending=False))



Missing values summary:
        Missing_Count  Missing_Percent
notes          266077             98.1
Medal          231333             85.3
Weight          62875             23.2
Height          60171             22.2
Age              9474              3.5
region            370              0.1


**Analysis of missing values**:

- **notes** (98.1%): It is fine, notes are only provided in rare cases
- **Medal** (85.3%): This is expected - most athletes do not win medals
- **Weight** (23.2%): Weight data is incomplete
- **Height** (22.2%): Height data is incomplete
- **Age** (3.5%): Personal information may not always be recorded
- **region** (0.1%): Some full named of the countries are not provided, which is not a problem

**Decision**: Keep missing values as-is since:
1. We won't be using `Weight` and `Height` columns as we don't have an objective to analyze weight and height.
2. The percentage of missing values for `Age` and `region` columns is relatively low.
3. Analysis functions like `.mean()` handle NaN automatically

For **Medal**, NaN represents "no medal" which is meaningful information.

#### Section 2: Data types

Now we need to ensure columns have appropriate data types for analysis.

We use `(data_frame.dtypes)` function to check current data types. 

In [58]:
print("Current data types:")
print(df.dtypes)

Current data types:
ID          int64
Name       object
Sex        object
Age       float64
Height    float64
Weight    float64
Team       object
NOC        object
Games      object
Year        int64
Season     object
City       object
Sport      object
Event      object
Medal      object
region     object
notes      object
dtype: object


Year column has an integer data type. 

We need to convert Year to datetime for time-based analysis of trends over 120 years of Olympic history.

Converting data type to datetime enables:
- Time series analysis
- Extraction of decade, century
- Chronological sorting and grouping
- Date-based filtering


In [66]:
df['Year_dt'] = pd.to_datetime(df['Year'], format="%Y")

print("Added Year_dt column:")
print(df[["Year", "Year_dt"]].head())
print(f"\nData type of year_dt: {df['Year_dt'].dtype}")

Added Year_dt column:
   Year    Year_dt
0  1992 1992-01-01
1  2012 2012-01-01
2  1920 1920-01-01
3  1900 1900-01-01
4  1988 1988-01-01

Data type of year_dt: datetime64[ns]


In [19]:
# TODO: Remove any duplicate entries
# there are 1385 duplicates, let's remove them
print(df.duplicated().sum())

1385


In [21]:
df_no_duplicates = df.drop_duplicates()
print(df_no_duplicates.head())

   ID                      Name Sex   Age  Height  Weight            Team  \
0   1                 A Dijiang   M  24.0   180.0    80.0           China   
1   2                  A Lamusi   M  23.0   170.0    60.0           China   
2   3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN         Denmark   
3   4      Edgar Lindenau Aabye   M  34.0     NaN     NaN  Denmark/Sweden   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   

   NOC        Games  Year  Season       City          Sport  \
0  CHN  1992 Summer  1992  Summer  Barcelona     Basketball   
1  CHN  2012 Summer  2012  Summer     London           Judo   
2  DEN  1920 Summer  1920  Summer  Antwerpen       Football   
3  DEN  1900 Summer  1900  Summer      Paris     Tug-Of-War   
4  NED  1988 Winter  1988  Winter    Calgary  Speed Skating   

                              Event Medal       region notes  
0       Basketball Men's Basketball   NaN        China   NaN  
1      Judo Men's Extra-Lightwei

In [13]:
# TODO: Anything else?

### Data wrangling

In [14]:
# TODO: Create at least one new column

In [15]:
# TODO: Combine at least two columns

In [16]:
# TODO: Extract data from multiple columns to create a new column

In [17]:
# TODO: Anything else?

## Data analysis

In [18]:
# TODO: Conduct data analysis

## Data visualisation

In [19]:
# TODO: Create at least two meaningful visualisations

In [20]:
# Example visualisation (you can modify or replace this)
# plt.figure(figsize=(10, 6))
# sns.countplot(x="Medal", data=df)
# plt.title("Distribution of Medals")
# plt.show()

In [21]:
# SUGGESTION: Use a small multiple  ???

## Data export

In [22]:
# TODO: Save the cleaned and wrangled DataFrame to a new CSV file

In [23]:
# TODO: Ensure that you can successfully load the exported CSV file

In [24]:
# TODO: Export visualizations as image files

## Conclusion

# Add your conclusion here [app. 300-500 words].

*Not sure where to add some challeanges we faced as a team:*

1 challenge - defining the aim and objectives of the project

2 challenge - Jaival faced challenge with data loading 
    this code didn't work at his machine (df = pd.read_csv("olympic_history/athlete_events.csv"))
    
3 I cannot upload Olympic data set on GitHub - the file is to large - more than 25 mb. 
Here is what AI replies:
"To upload large files to GitHub, since your data folder is more than 25 MB, you cannot upload it via the GitHub web interface due to a 25 MB size limit per file on browser uploads. However, the Git limit for individual files pushed via Git commands is 100 MB. For files larger than 100 MB, GitHub blocks the push entirely.
To upload files larger than 25 MB but smaller than 100 MB, you should use Git command line tools or Git clients like GitHub Desktop:"

After asking AI and Jaival I realised I need to install Desktop version of Git in order to upload our Olympic data sets on GitHub. 


## References

1. Reference 1
2. Reference 2
3. Reference 3
4. ...

*(References must be complete and contain a link to the original source; usually a DOI link.)*

## Appendices

### Appendix A

Displaying the first 5 rows of athlete_events.csv to familiarise ourselves with the raw data it contains

In [25]:
print(athlete_events.head())

   ID                      Name Sex   Age  Height  Weight            Team  \
0   1                 A Dijiang   M  24.0   180.0    80.0           China   
1   2                  A Lamusi   M  23.0   170.0    60.0           China   
2   3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN         Denmark   
3   4      Edgar Lindenau Aabye   M  34.0     NaN     NaN  Denmark/Sweden   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   

   NOC        Games  Year  Season       City          Sport  \
0  CHN  1992 Summer  1992  Summer  Barcelona     Basketball   
1  CHN  2012 Summer  2012  Summer     London           Judo   
2  DEN  1920 Summer  1920  Summer  Antwerpen       Football   
3  DEN  1900 Summer  1900  Summer      Paris     Tug-Of-War   
4  NED  1988 Winter  1988  Winter    Calgary  Speed Skating   

                              Event Medal  
0       Basketball Men's Basketball   NaN  
1      Judo Men's Extra-Lightweight   NaN  
2           Football Men's

### Appendix B

Displaying the first 5 rows of noc_regions.csv to familiarise ourselves with the raw data in it

In [26]:
print(noc_regions.head())

   NOC       region                 notes
0  AFG  Afghanistan                   NaN
1  AHO      Curacao  Netherlands Antilles
2  ALB      Albania                   NaN
3  ALG      Algeria                   NaN
4  AND      Andorra                   NaN
